### PR TITLE
feat: [😭] Firearm `AmmoModules` Api

### PR DESCRIPTION
**Pull request by IRacle1**
## Description
**Describe the changes** 

*inspired by* #255 

I was thinking about more flexible logic for setting ammo. My first idea was to make better abstractions of `Barrel` and `Primary` magazines(that what i did in that pr firstly). I liked that idea, and that gave me a possibility to add more specific api for specific magazines (`Rotate` for revolvers, or bolt locking for automatic firearms as example). But there is 2 questions about ammo setters and for modules in general:
 - there was a discussion in 255 pr, [taptap](https://github.com/ExMod-Team/EXILED/pull/255#discussion_r1863846886) , for giving support to more than one primary/barrel magazines at ones. Ive thought about it later, and Ive decided in this case to move away from several magazines abstraction in favour of easier system of maximum 2 magazines (barrel and primary). But if adding custom modules turns to be real and possible, were gonna have to look at this again.
 - main question is an about ammo setter at all, do we generally need such complicated `Ammo`/`MaxAmmo` setters? When i finished with main part of api, and started thinking about total `Ammo` setter. And I really dont think an absolute/total ammo setter like before 14.0 is needed. This setter has an overly complex background and logic. Even if we position Exiled as a highlevel api, I still dont think its necessary to make a single setter for barrel and magazine ammos. But anyway that my honest shizo opinion says. Really wait for feedback😘

*oh, and i added RotatingRevolver event😎*

**What is the current behavior?** (You can also link to an open issue here)
😥

**What is the new behavior?** (if this is a feature change)
😎

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
😓

**Other information**:
🥹
<br />

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentations
<br />

## Submission checklist

- [x] I have checked the project can be compiled
- [x] I have tested my changes and it worked as expected

### Patches (if there are any changes related to Harmony patches)
- [x] I have checked no IL patching errors in the console

## Main Axioms that i used for api logic, or tldr😎
- `Barrel` and `Normal` magazines should be clearly segregated for the convenience of people
- In those abstraction `Ammo` setters should be available, despite the lack of this abstraction in the basegame.
- As i think now, the best way will to create `BarrelAmmo` `MagazineAmmo` full properties, and make `TotalAmmo` only with getter

### CODE REVIEW

**Comment by IRacle1** [comment]
All ammoModules🥹
```yaml
firearm: GunCOM15
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunCOM18
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunE11SR
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunCrossvec
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunFSP9
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunLogicer
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunRevolver
 - primary: CylinderAmmoModule
firearm: GunAK
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunShotgun
 - barrel: PumpActionModule
 - primary: MagazineModule
firearm: GunCom45
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: ParticleDisruptor
 - primary: MagazineModule
firearm: GunFRMG0
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunA7
 - barrel: AutomaticActionModule
 - primary: MagazineModule
```

**Comment by IRacle1** [comment]
Anyway until we come to an agreement on that Ill leave it to the draft

**Comment by Misaka-ZeroTwo** [comment]
I like the idea of separating the barrel and the magazine. 

It will be way too complex to handle them all in one place. We would deal with more edge cases and a much higher chance that something just doesn't work.

**Comment by VALERA771** [comment]
Lgtm. For some people may be too complex to use but game the game is also is not too easy to understand. So if people don't need to use a lot of classes to get a ammo and and if we don't have a 70 line ammo setter I agree

**Comment by VALERA771** [comment]
Also waiting for these guys opinions

File: EXILED/Exiled.API/Features/Items/Firearm.cs [hunk 1]
```diff
@@ -105,24 +123,97 @@ public static IReadOnlyDictionary<Player, Dictionary<FirearmType, AttachmentIden
...
+        /// not working for Revolver and ParticleDisruptor.
+        /// </remarks>
+        public int MaxBarrelAmmo
         {
-            get => (Base.Modules[Array.IndexOf(Base.Modules, typeof(MagazineModule))] as MagazineModule).AmmoMax;
-            set => (Base.Modules[Array.IndexOf(Base.Modules, typeof(MagazineModule))] as MagazineModule)._defaultCapacity = value; // Synced?
+            get
+            {
+                return BarrelMagazine?.MaxAmmo ?? 0;
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/Firearm.cs:193]: Inlined this

File: EXILED/Exiled.API/Features/Items/Firearm.cs [hunk 2]
```diff
@@ -105,24 +123,97 @@ public static IReadOnlyDictionary<Player, Dictionary<FirearmType, AttachmentIden
...
+                if (BarrelMagazine != null)
+                    BarrelMagazine.MaxAmmo = value;
+            }
         }
 
+        /// <summary>
+        /// Gets the total amount of ammo in the firearm.
+        /// </summary>
+        public int TotalMaxAmmo =>
+            Base.GetTotalMaxAmmo();
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/Firearm.cs:206]: Inlined these

File: EXILED/Exiled.API/Features/Items/Firearm.cs [hunk 3]
```diff
@@ -105,24 +123,97 @@ public static IReadOnlyDictionary<Player, Dictionary<FirearmType, AttachmentIden
...
+
+        /// <summary>
+        /// Gets or sets a ammo drain per shoot.
+        /// </summary>
+        /// <remarks>
+        /// Always <see langword="1"/> by default.
+        /// Applied on a high layer nether basegame ammo controllers.
+        /// </remarks>
+        public int AmmoDrain { get; set; } = 2;
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/Firearm.cs:215]: My brain didn't get what these remarks mean and why 2 is needed

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        /// <inheritdoc/>
+        public override Firearm Firearm { get; }
+
+        /// <inheritdoc/>
+        public override int Ammo
+        {
+            get
+            {
+                return PumpBarrel.SyncChambered;
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:50]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+        /// <summary>
+        /// Gets or sets an amount of bullets, that pump module will try to shot.
+        /// </summary>
+        public int CockedAmmo
+        {
+            get
+            {
+                return PumpBarrel.SyncCocked;
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:68]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        {
+            get
+            {
+                return PumpBarrel._numberOfBarrels;
+            }
+
+            set
+            {
+                PumpBarrel._numberOfBarrels = Mathf.Max(value, 0);
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:89]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        /// <inheritdoc/>
+        public override bool IsCocked
+        {
+            get
+            {
+                return PumpBarrel.SyncCocked > 0;
+            }
+
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:99]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                PumpBarrel.SyncCocked = MaxAmmo;
+                Resync();
+            }
+        }
+
+        /// <inheritdoc/>
+        public override void Resync()
+        {
+            PumpBarrel.ServerResync();
+        }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:111]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        {
+            int oldAmmo = Ammo;
+            if (useBorders)
+            {
+                Ammo = Mathf.Clamp(Ammo + delta, 0, MaxAmmo);
+            }
+            else
+            {
+                Ammo += delta;
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs:96]: Can be inlined

File: EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            return Ammo - oldAmmo;
+        }
+
+        /// <summary>
+        /// Fills current <see cref="Ammo"/> to <see cref="MaxAmmo"/>.
+        /// </summary>
+        public void Fill()
+        {
+            Ammo = MaxAmmo;
+        }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs:107]: Inlined these

**Review by louis1706 — CHANGES_REQUESTED** [review]
I am lazzy to finish just inlined everything than can be inlined

File: EXILED/Exiled.API/Features/Items/Firearm.cs (hunk 3 above)
Comment (IRacle1) [reply EXILED/Exiled.API/Features/Items/Firearm.cs:215]: Oh, i forgot to set it for 1 by default, i used `2` for testing😭

File: EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs (hunk 9 above)
Comment (IRacle1) [reply EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs:96]: How🤨

Comment (louis1706) [reply EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs:96]:
```suggestion
            ammo = useBorders ? Mathf.Clamp(Ammo + delta, 0, MaxAmmo) : Ammo + delta;
```

**Comment by VALERA771** [comment]
Is this PR is ready for reviews and tests? I think it's one the best one from other PRs which are fixing firearm properties

**Comment by VALERA771** [comment]
Some things (a lot) can be inlined. But generally looks good if you've tested magazine API

File: EXILED/Exiled.API/Features/Items/Firearm.cs [hunk 11]
```diff
@@ -105,24 +123,95 @@ public static IReadOnlyDictionary<Player, Dictionary<FirearmType, AttachmentIden
...
+                if (BarrelMagazine != null)
+                    BarrelMagazine.Ammo = value;
+            }
         }
 
+        /// <summary>
+        /// Gets the total amount of ammo in the firearm.
+        /// </summary>
+        public int TotalAmmo =>
+            Base.GetTotalStoredAmmo();
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/Firearm.cs:168]:
```suggestion
        public int TotalAmmo => Base.GetTotalStoredAmmo();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/AutomaticBarrelMagazine.cs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        /// Will be ranged between <see langword="0"/> and <see langword="16"/> due basegame logic.
+        /// </remarks>
+        public override int MaxAmmo
+        {
+            get => AutomaticBarrel.ChamberSize;
+
+            set
+            {
+                AutomaticBarrel.ChamberSize = Mathf.Clamp(value, 0, 16);
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/AutomaticBarrelMagazine.cs:66]:
```suggestion
            set => AutomaticBarrel.ChamberSize = Mathf.Clamp(value, 0, 16);
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        /// <inheritdoc/>
+        public override int ConstantMaxAmmo => MagazineModule._defaultCapacity;
+
+        /// <inheritdoc/>
+        public override AmmoType AmmoType
+        {
+            get
+            {
+                return Magazine.AmmoType.GetAmmoType();
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:58]:
```suggestion
            get => Magazine.AmmoType.GetAmmoType();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+            set
+            {
+                MagazineModule._ammoType = value.GetItemType();
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:63]:
```suggestion
            set => MagazineModule._ammoType = value.GetItemType();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+        /// <summary>
+        /// Gets or sets a value indicating whether magazine is inserted.
+        /// </summary>
+        public bool MagazineInserted
+        {
+            get
+            {
+                return MagazineModule.MagazineInserted;
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:74]:
```suggestion
            get => MagazineModule.MagazineInserted;
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        /// Removes magazine from current <see cref="Exiled.API.Features.Items.Firearm"/>.
+        /// </summary>
+        /// <remarks>
+        /// Affects on actual ammo count.
+        /// Removes all ammo from magazine.
+        /// </remarks>
+        public void RemoveMagazine()
+        {
+            MagazineModule.ServerRemoveMagazine();
+        }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:93]:
```suggestion
        public void RemoveMagazine() => MagazineModule.ServerRemoveMagazine();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+        /// <summary>
+        /// Inserts current magazine from current <see cref="Exiled.API.Features.Items.Firearm"/>.
+        /// </summary>
+        public void InsertMagazine()
+        {
+            MagazineModule.ServerInsertEmptyMagazine();
+        }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:101]:
```suggestion
        public void InsertMagazine() => MagazineModule.ServerInsertEmptyMagazine();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+        /// <inheritdoc/>
+        public override void Resync()
+        {
+            MagazineModule.ServerResyncData();
+        }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:107]:
```suggestion
        public override void Resync() => MagazineModule.ServerResyncData();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/PrimaryMagazine.cs [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+        public IPrimaryAmmoContainerModule Magazine { get; }
+
+        /// <inheritdoc/>
+        public override int MaxAmmo
+        {
+            get
+            {
+                return Magazine.AmmoMax;
+            }
+        }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/PrimaryMagazine.cs:44]:
```suggestion
        public override int MaxAmmo => Magazine.AmmoMax;
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/PrimaryMagazine.cs [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+        /// </summary>
+        public abstract int ConstantMaxAmmo { get; }
+
+        /// <inheritdoc/>
+        public override int Ammo
+        {
+            get
+            {
+                return Magazine.AmmoStored;
+            }
```
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/PrimaryMagazine.cs:57]:
```suggestion
            get => Magazine.AmmoStored;
```

File: EXILED/Exiled.Events/EventArgs/Player/RotatingRevolverEventArgs.cs [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+        public int Rotations { get; set; }
+
+        /// <summary>
+        /// Gets a value indicating whether the rotation will have an effect.
+        /// </summary>
+        /// <remarks>
+        /// checks rotations and chambers counts equality by mod of chambers counts. <code>Rotations % Chambers.Length == 0</code>
+        /// </remarks>
+        public bool HasEffect =>
+            Firearm.PrimaryMagazine.MaxAmmo % Rotations == 0;
```
Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Player/RotatingRevolverEventArgs.cs:54]:
```suggestion
        public bool HasEffect => Firearm.PrimaryMagazine.MaxAmmo % Rotations == 0;
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+            yield return new CodeInstruction(OpCodes.Dup);
+            yield return new CodeInstruction(OpCodes.Stloc_S, firearm.LocalIndex);
+            yield return new CodeInstruction(OpCodes.Brfalse_S, continueLabel);
+
+            yield return new CodeInstruction(OpCodes.Ldloc_S, firearm.LocalIndex);
+            yield return new CodeInstruction(OpCodes.Callvirt, PropertyGetter(typeof(API.Features.Items.Firearm), nameof(API.Features.Items.Firearm.AmmoDrain)));
+            yield return new CodeInstruction(OpCodes.Stloc_S, ammoDrain.LocalIndex);
+
+            yield return new CodeInstruction(OpCodes.Nop).WithLabels(continueLabel);
+        }
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:62]: Is it possible to comment this part of code ? to explain more what it's does

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+
+            LocalBuilder firearm = generator.DeclareLocal(typeof(API.Features.Items.Firearm));
+            LocalBuilder ammoDrain = generator.DeclareLocal(typeof(int));
+
+            Label cnt = generator.DefineLabel();
+
+            // insert default instructions for getting ammo drain
+            newInstructions.InsertRange(
+                0,
+                GetInstructions(firearm, ammoDrain, cnt));
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:79]:
```suggestion
            newInstructions.InsertRange(0, GetInstructions(firearm, ammoDrain, cnt));
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+
+            // divide max ammo by AmmoDrain
+            // we cant cock more ammo than we have in primary magazine, so we limit it due ammo drain
+            newInstructions.InsertRange(
+                index,
+                new CodeInstruction[]
+                {
+                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
+                    new(OpCodes.Div),
+                });
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:92]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Div),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 25]
```diff
@@ -0,0 +1,264 @@
...
+            index = newInstructions.FindIndex(i => i.Calls(Method(typeof(IPrimaryAmmoContainerModule), nameof(IPrimaryAmmoContainerModule.ServerModifyAmmo)))) + offset;
+
+            // multiply ammo that are removed from primary magazine, actuall logic of ammo drain
+            newInstructions.InsertRange(
+                index,
+                new CodeInstruction[]
+                {
+                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
+                    new(OpCodes.Mul),
+                });
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:104]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Mul),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 26]
```diff
@@ -0,0 +1,264 @@
...
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(codeInstructions);
+
+            LocalBuilder firearm = generator.DeclareLocal(typeof(API.Features.Items.Firearm));
+            LocalBuilder ammoDrain = generator.DeclareLocal(typeof(int));
+
+            // insert default instructions for getting ammo drain
+            Label cnt = generator.DefineLabel();
+            newInstructions.InsertRange(
+                0,
+                GetInstructions(firearm, ammoDrain, cnt));
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:126]:
```suggestion
            newInstructions.InsertRange(0, GetInstructions(firearm, ammoDrain, cnt));
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 27]
```diff
@@ -0,0 +1,264 @@
...
+
+            // divide max ammo by AmmoDrain
+            // we cant cock more ammo than we have in primary magazine, so we limit it due ammo drain
+            newInstructions.InsertRange(
+                index,
+                new CodeInstruction[]
+                {
+                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
+                    new(OpCodes.Div),
+                });
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:139]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Div),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 28]
```diff
@@ -0,0 +1,264 @@
...
+            index = newInstructions.FindIndex(i => i.Calls(Method(typeof(IPrimaryAmmoContainerModule), nameof(IPrimaryAmmoContainerModule.ServerModifyAmmo)))) + offset;
+
+            // multiply ammo that are removed from primary magazine, actuall logic of ammo drain
+            newInstructions.InsertRange(
+                index,
+                new CodeInstruction[]
+                {
+                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
+                    new(OpCodes.Mul),
+                });
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:151]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Mul),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+
+            LocalBuilder firearm = generator.DeclareLocal(typeof(API.Features.Items.Firearm));
+            LocalBuilder ammoDrain = generator.DeclareLocal(typeof(int));
+
+            Label cnt = generator.DefineLabel();
+
+            // insert default instructions for getting ammo drain
+            newInstructions.InsertRange(
+                0,
+                AmmoDrainAutomatic.GetInstructions(firearm, ammoDrain, cnt));
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:178]:
```suggestion
            newInstructions.InsertRange(0, AmmoDrainAutomatic.GetInstructions(firearm, ammoDrain, cnt));
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 30]
```diff
@@ -0,0 +1,264 @@
...
+
+            // divide max ammo by AmmoDrain
+            // we cant cock more ammo than we have in primary magazine, so we limit it due ammo drain
+            newInstructions.InsertRange(
+                index,
+                new CodeInstruction[]
+                {
+                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
+                    new(OpCodes.Div),
+                });
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:191]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Div),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 31]
```diff
@@ -0,0 +1,264 @@
...
+            index = newInstructions.FindIndex(i => i.Calls(Method(typeof(IPrimaryAmmoContainerModule), nameof(IPrimaryAmmoContainerModule.ServerModifyAmmo)))) + offset;
+
+            // multiply ammo that are removed from primary magazine, actuall logic of ammo drain
+            newInstructions.InsertRange(
+                index,
+                new CodeInstruction[]
+                {
+                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
+                    new(OpCodes.Mul),
+                });
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:203]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Mul),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 32]
```diff
@@ -0,0 +1,264 @@
...
+
+            LocalBuilder firearm = generator.DeclareLocal(typeof(API.Features.Items.Firearm));
+            LocalBuilder ammoDrain = generator.DeclareLocal(typeof(int));
+
+            Label cnt = generator.DefineLabel();
+
+            // insert default instructions for getting ammo drain
+            newInstructions.InsertRange(
+                0,
+                AmmoDrainAutomatic.GetInstructions(firearm, ammoDrain, cnt));
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:230]:
```suggestion
            newInstructions.InsertRange(0, AmmoDrainAutomatic.GetInstructions(firearm, ammoDrain, cnt));
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 33]
```diff
@@ -0,0 +1,264 @@
...
+            // multiply max ammo by AmmoDrain
+            // there is little different logic for revolver, for other weapons we can patch methods for removing ammo after shooting/pumping, but for revolver we only can patch method for reload
+            // so our actions are inverted, multiply max ammo limit to select more ammo due AmmoDrain
+            newInstructions.InsertRange(
+                index,
+                new CodeInstruction[]
+                {
+                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
+                    new(OpCodes.Mul),
+                });
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:244]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Mul),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs [hunk 34]
```diff
@@ -0,0 +1,264 @@
...
+            index = newInstructions.FindIndex(i => i.Calls(Method(typeof(IPrimaryAmmoContainerModule), nameof(IPrimaryAmmoContainerModule.ServerModifyAmmo)))) + offset;
+
+            // and divide ammo that are inserting in magazine, to implement AmmoDrain
+            newInstructions.InsertRange(
+                index,
+                new CodeInstruction[]
+                {
+                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
+                    new(OpCodes.Div),
+                });
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:256]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Div),
                });
```

**Review by louis1706 — APPROVED** [review]
It's all fine to merge when this will be completed (it's can be done after it's merge if needed)

the only important it's for AmmoDrain.cs file about comment on the GetInstructions method

**Review by BoltonDev — APPROVED** [review]

**Comment by BonesMgee** [comment]
Epic 🧀

**Pull request merged**